### PR TITLE
 End-of-form navigation manual linking advanced mode

### PR DESCRIPTION
**Pull request by esoergel**
## Product Description
Adds an "advanced mode" feature flag for end of form navigation manual linking.

Without the feature flag:
image

With the feature flag:
image

## Technical Summary
https://dimagi-dev.atlassian.net/browse/USH-3966

Note: This reorganizes https://github.com/dimagi/commcare-hq/pull/33950, which got pretty messy.  It also removes the changes to `get_form_datums`, which ended up not being useful.  Commits should be much easier to review now.

EoF navigation involves populating the stack to navigate through the app to the desired target.  Some forms and modules expect certain variables (like case IDs) to be available in the session.  Automatic linking attempts to align existing session variables with those needed to navigate to the target.  Where that isn't possible, the app builder must provide an xpath expression for each of these datum IDs.  This manual linking is also sometimes used when automatic linking doesn't do what's desired.

However, we can't always reliably get the set of datum IDs needed to get to a target without going through the full build process, especially when child and shadow modules are involved, as the datum IDs are modified at various stages in the build process.

This PR introduces an "advanced mode" feature flag for manual linking that essentially gives up trying to predict the IDs needed, and requires the user to specify all sets of IDs and xpath expressions needed to navigate to the endpoint.  This will likely be a pain to use, though build validation should catch errors here.

Depending on how this is used, we may consider an extension of this work where advanced mode is used only on a case-by-case basis, so app builders could decide whether to enable it per module, and otherwise use the existing manual linking UI (we'd need to ensure switching between the two doesn't break anything).

In testing this workflow out, we ran into a couple suite generation bugs which I've also addressed here in the last couple commits.

## Feature Flag
This adds a new feature flag: 
FORM_LINK_ADVANCED_MODE - USH: Form linking advanced mode

### Safety story
Testing on staging.  The UI itself is pretty small, it uses an existing widget for specifying key/value pairs.  Everything should be sufficiently hidden behind the feature flag that I'm not too concerned about affecting projects that don't enable the FF.

### Automated test coverage

none

### QA Plan

none

### Rollback instructions
 
- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

File: corehq/apps/app_manager/suite_xml/post_process/workflow.py [hunk 1]
```diff
@@ -272,16 +272,14 @@ def _get_datums_matched_to_manual_values(target_frame_elements, manual_values, f
     """
     manual_values_by_name = {datum.name: datum.xpath for datum in manual_values}
     for child in target_frame_elements:
-        if not isinstance(child, WorkflowSessionMeta) or not child.requires_selection:
+        if child.id in manual_values_by_name:
```
Comment (MartinRiese) [inline corehq/apps/app_manager/suite_xml/post_process/workflow.py:275]: That is a lot more readable. Thanks for cleaning up.

**Review by MartinRiese — APPROVED** [review]
Looks good.

**Comment by esoergel** [comment]
aside: I'm considering an alternative UI approach, where we extend the existing UI with an "Add Datum" button like this:
image

I don't love it as-is though, since we'd also probably want a "delete" button, and clicking "refresh" resets the values.  I don't think it'd be hard to improve this, though maybe we should do a design pass before undertaking such a thing, especially if it'd be GA.

File: corehq/apps/app_manager/static/app_manager/js/forms/form_workflow.js [hunk 2]
```diff
@@ -192,6 +194,26 @@ hqDefine('app_manager/js/forms/form_workflow', function () {
...
         });
 
+        if (self.advancedMode) {
+            self.advancedModeDatumsElement = uiElementKeyValueList.new(
+                String(Math.random()).slice(2),  // random id
+                gettext("Manual Linking Datums"),
+                gettext("Set datums required to navigate to the selected form or menu"),
+                {key: gettext("Datum ID"), value: gettext("XPath Expression")}  // placeholders
+            );
+            let datumsDict = {};
```
[the remark this answers is not in the document]
Comment (esoergel) [reply corehq/apps/app_manager/static/app_manager/js/forms/form_workflow.js]: Ah right thanks - I think we had the same conversation on the old PR

**Comment by MartinRiese** [comment]
I am a bit torn about the UI too. But I am not sure having all of the config on the page itself is great either. Makes it harder to find stuff. Either way. I think we should use it first and see if we need any other configuration that we are not thinking about and then do a design.

**Review by MartinRiese — APPROVED** [review]

**Pull request merged**